airbyte-ci/connectors/pipelines/pipelines/helpers/changelog.py

```diff
@@ -0,0 +1,117 @@
+from pathlib import Path
+
+import semver
+from typing_extensions import Set
+
+
+class Changelog:
+    def __init__(self, changelog_file_path: Path) -> None:
+        self.changelog_file_path = changelog_file_path
+        if not changelog_file_path.exists():
+            raise Exception(f"File {changelog_file_path} does not exist")
+        if not changelog_file_path.is_file():
+            raise Exception(f"File {changelog_file_path} is not a regular file")
+        self.original_markdown_lines = changelog_file_path.read_text().splitlines()
+        self.__parse_markdown()
+        self.new_entries: Set[ChangelogEntry] = set()
+
+    def add_entry(self, version: semver.Version, date: datetime.date, pull_request_number: int, comment: str) -> None:
+        self.new_entries.add(ChangelogEntry(date, version, pull_request_number, comment))
+
+    def to_markdown(self) -> str:
+        all_entries = set(self.old_entries.union(self.new_entries))
+        sorted_entries = sorted(
+            sorted(
+                sorted(sorted(all_entries, key=attrgetter("comment"), reverse=True), key=attrgetter("pr_number"), reverse=True),
+                key=attrgetter("date"),
+                reverse=True,
+            ),
+            key=attrgetter("version"),
+            reverse=True,
+        )
+        new_lines = (
+            self.original_markdown_lines[: self.changelog_entries_start_line_index]
+            + [line.to_str() for line in sorted_entries]
+            + self.original_markdown_lines[self.last_changelog_line_index + 1 :]
+        )
+        return "\n".join(new_lines) + "\n"
+
+    def __parse_markdown(self) -> None:
+        changelog_entry_re = (
+            "^\\| *(?P<version>[0-9]+\\.[0-9+]+\\.[0-9]+?) *\\| *"
+            + "(?P<day>[0-9]{4}-[0-9]{2}-[0-9]{2}) *\\| *"
+            + "\\[?(?P<pr_number1>[0-9]*)\\]? ?\\(https://github.com/airbytehq/airbyte/pull/(?P<pr_number2>[0-9]*)\\) *\\| *"
+            + "(?P<comment>[^ ].*[^ ]) *\\| *$"
+        )
+        changelog_header_line_index = -1
+        changelog_line_enumerator = enumerate(self.original_markdown_lines)
+        for line_index, line in changelog_line_enumerator:
+            if re.search(r"\| *Version *\| *Date *\| *Pull Request *\| *Subject *\|", line):
+                changelog_header_line_index = line_index
+                break
+        if changelog_header_line_index == -1:
+            raise Exception("Could not find the changelog section table in the documentation file.")
+        if self.original_markdown_lines[changelog_header_line_index - 1] != "":
+            raise Exception("Found changelog section table in the documentation file at line but there is not blank line before it.")
+        if not re.search(r"(\|-*){4}\|", next(changelog_line_enumerator)[1]):
+            raise Exception("The changelog table in the documentation file is missing the header delimiter.")
+        self.changelog_entries_start_line_index = changelog_header_line_index + 2
```

Comment:
I did not see `self.changelog_entries_start_line_index` being initialized at the constructor level or at the class variable level? Is it on purpose?

Reply:
yes, the parse_markdown is responsible for setting it.